content/advent-2019/safe-use-of-unsafe-pointer.md

```diff
@@ -0,0 +1,325 @@
+that the first 8 bytes of the array contain our `uint64` data in its native
+endian format (little endian on my machine):
+
+```
+$ go run main.go
+[]byte{0xff, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}
+```
+
+However, it turns out there is an issue with this program as well. If we attempt
+to run the program on Go tip with the `checkptr` debug flag, we will see:
+
+```
+$ gotip run -gcflags=all=-d=checkptr main.go 
+panic: runtime error: unsafe pointer conversion
+
+goroutine 1 [running]:
+main.main()
+        /home/matt/src/github.com/mdlayher/example/main.go:17 +0x60
+exit status 2
+```
+
+This check is still quite new and as such does not provide much information
+beyond the "unsafe pointer conversion" panic message and a stack trace. But the
+stack trace does at least provide a hint that line 17 is suspect.
+
+By casting a smaller structure into a larger one, we enable reading arbitrary
+memory beyond the end of the smaller structure's data! This is another way that
+careless use of `unsafe` could result in a security vulnerability in your
+application.
+
+In order to perform this operation safely, we have to make sure that we
+initialize the "union" structure first before copying data into it, so we can
+ensure that arbitrary memory is not accessed:
+
+```go
+package main
+
+import (
+    "fmt"
+    "unsafe"
+)
+
+// one is a typed Go structure containing structured data to pass to the kernel.
+type one struct{ v uint64 }
+
+// two mimics a C union type which passes a blob of data to the kernel.
+type two struct{ b [32]byte }
+
+// newTwo safely produces a two structure from an input one.
+func newTwo(in one) *two {
+    // Initialize out and its array.
+    var out two
+
+    // Explicitly copy the contents of in into out by casting both into byte
+    // arrays and then slicing the arrays. This will produce the correct packed
+    // union structure, without relying on unsafe casting to a smaller type of a
+    // larger type.
+    copy(
+        (*(*[unsafe.Sizeof(two{})]byte)(unsafe.Pointer(&out)))[:],
+        (*(*[unsafe.Sizeof(one{})]byte)(unsafe.Pointer(&in)))[:],
+    )
+
+    return &out
+}
+
+func main() {
+    // All is well! The two structure is appropriately initialized.
+    out := newTwo(one{v: 0xff})
+
+    fmt.Printf("%#v\n", out.b[:8])
+}
+```
+
+We can now run our updated program with the same flags as before, and we will
+see that the issue has been resolved:
+
+```
+$ gotip run -gcflags=all=-d=checkptr main.go 
+[]byte{0xff, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}
+```
+
+By removing the slicing operation from the `fmt.Printf` call, we can verify that
+the remainder of the byte array has been initialized to `0` bytes:
+
+```
+[32]uint8{
+	0xff, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+	0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+	0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+	0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+}
+```
+
+This is a very easy mistake to make, and in fact, I recently had to fix this
```

Comment:
Maybe start with this story at the top of the blog post, it'll give some motivation to read it.

Reply:
Sure, I will add a bit of a teaser.